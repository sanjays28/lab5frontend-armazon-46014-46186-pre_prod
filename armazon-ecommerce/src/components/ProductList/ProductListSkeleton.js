import React from 'react';
import { Grid, Container, Box, Skeleton, Card } from '@mui/material';
import { styled } from '@mui/material/styles';

const StyledCard = styled(Card)(({ theme }) => ({
  height: '100%',
  display: 'flex',
  flexDirection: 'column',
}));

const ProductCardSkeleton = () => (
  <StyledCard>
    <Skeleton
      variant="rectangular"
      sx={{
        paddingTop: '56.25%', // 16:9 aspect ratio
        backgroundSize: 'contain',
      }}
      animation="wave"
    />
    <Box sx={{ p: 2, flexGrow: 1 }}>
      <Skeleton variant="text" width="80%" height={32} sx={{ mb: 1 }} animation="wave" />
      <Skeleton variant="text" width="100%" height={20} animation="wave" />
      <Skeleton variant="text" width="100%" height={20} animation="wave" />
      <Skeleton variant="text" width="40%" height={32} sx={{ mt: 2 }} animation="wave" />
    </Box>
    <Box sx={{ p: 2, pt: 0 }}>
      <Skeleton variant="rectangular" height={36} width="100%" animation="wave" />
    </Box>
  </StyledCard>
);

// PUBLIC_INTERFACE
const ProductListSkeleton = () => {
  return (
    <Box sx={{ flexGrow: 1, py: 4 }}>
      <Container maxWidth="lg">
        <Box sx={{ mb: 3 }}>
          <Skeleton variant="rectangular" height={56} animation="wave" />
        </Box>
        <Grid container spacing={3}>
          {[...Array(8)].map((_, index) => (
            <Grid item key={index} xs={12} sm={6} md={4} lg={3}>
              <ProductCardSkeleton />
            </Grid>
          ))}
        </Grid>
      </Container>
    </Box>
  );
};

export default ProductListSkeleton;