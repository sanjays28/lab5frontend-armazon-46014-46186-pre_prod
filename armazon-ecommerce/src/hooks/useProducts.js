import { useState, useEffect } from 'react';
import { getProducts, searchProducts } from '../services/productService';

// PUBLIC_INTERFACE
const useProducts = (searchQuery = '') => {
  const [products, setProducts] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    const fetchProducts = async () => {
      setLoading(true);
      try {
        const result = searchQuery
          ? await searchProducts(searchQuery)
          : await getProducts();

        if (result.error) {
          throw new Error(result.error);
        }

        setProducts(result.data);
        setError(null);
      } catch (err) {
        setError(err.message);
        setProducts([]);
      } finally {
        setLoading(false);
      }
    };

    fetchProducts();
  }, [searchQuery]);

  return {
    products,
    loading,
    error,
  };
};

export default useProducts;