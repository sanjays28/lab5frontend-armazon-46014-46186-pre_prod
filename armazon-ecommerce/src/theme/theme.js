import { createTheme } from '@mui/material/styles';

// PUBLIC_INTERFACE
const theme = createTheme({
  palette: {
    primary: {
      main: '#1976d2',
    },
    secondary: {
      main: '#dc004e',
    },
  },
  typography: {
    fontFamily: [
      '-apple-system',
      'BlinkMacSystemFont',
      '"Segoe UI"',
      'Roboto',
      '"Helvetica Neue"',
      'Arial',
      'sans-serif',
    ].join(','),
  },
});

export default theme;