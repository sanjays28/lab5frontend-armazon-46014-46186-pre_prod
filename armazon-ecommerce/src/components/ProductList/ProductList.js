import React, { useState } from 'react';
import { Grid, Container, Box, Button, Alert } from '@mui/material';
import ProductCard from '../ProductCard/ProductCard';
import SearchBar from '../SearchBar/SearchBar';
import ProductListSkeleton from './ProductListSkeleton';
import useProducts from '../../hooks/useProducts';

// PUBLIC_INTERFACE
const ProductList = ({ onAddToCart }) => {
  const [searchQuery, setSearchQuery] = useState('');
  const { products, loading, error } = useProducts(searchQuery);

  const handleRetry = () => {
    setSearchQuery(searchQuery); // This will trigger a re-fetch
  };

  const renderContent = () => {
    if (loading) {
      return <ProductListSkeleton data-testid="loading-spinner" />;
    }

    if (error) {
      return (
        <Box sx={{ py: 4, textAlign: 'center' }}>
          <Alert 
            severity="error" 
            action={
              <Button color="inherit" size="small" onClick={handleRetry}>
                RETRY
              </Button>
            }
            data-testid="error-message"
          >
            {error}
          </Alert>
        </Box>
      );
    }

    if (!products || !products.length) {
      return (
        <Box sx={{ py: 4, textAlign: 'center' }}>
          <Alert severity="info" data-testid="no-products-message">
            No products found. Try adjusting your search criteria.
          </Alert>
        </Box>
      );
    }

    return (
      <Grid container spacing={3}>
        {products.map((product) => (
          <Grid item key={product.id} xs={12} sm={6} md={4} lg={3}>
            <ProductCard 
              product={product}
              onAddToCart={onAddToCart}
            />
          </Grid>
        ))}
      </Grid>
    );
  };
  return (
    <Box sx={{ flexGrow: 1, py: 4 }}>
      <Container maxWidth="lg">
        <Box sx={{ mb: 3 }}>
          <SearchBar 
            onSearch={setSearchQuery} 
            isLoading={loading} 
            disabled={loading}
          />
        </Box>
        {renderContent()}
      </Container>
    </Box>
  );
};

export default ProductList;
