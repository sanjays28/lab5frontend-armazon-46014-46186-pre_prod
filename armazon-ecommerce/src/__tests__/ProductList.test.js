import React from 'react';
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '@mui/material/styles';
import ProductList from '../components/ProductList/ProductList';
import theme from '../theme/theme';

// Helper function to wait for Material-UI animations
const waitForMuiAnimation = () => new Promise(resolve => setTimeout(resolve, 0));

// Mock the custom hook
jest.mock('../hooks/useProducts', () => ({
  __esModule: true,
  default: jest.fn()
}));

// Mock the child components
jest.mock('../components/ProductCard/ProductCard', () => {
  return function MockProductCard({ product }) {
    return (
      <div data-testid={`product-card-${product.id}`} className="product-card" key={product.id}>
        <div data-testid={`product-name-${product.id}`}>{product.name}</div>
        <div data-testid={`product-price-${product.id}`}>${product.price}</div>
      </div>
    );
  };
});

jest.mock('../components/SearchBar/SearchBar', () => {
  return function MockSearchBar({ onSearch, isLoading }) {
    return (
      <div data-testid="search-bar" className="search-bar">
        <input
          type="text"
          data-testid="search-input"
          placeholder="Search products..."
          onChange={(e) => onSearch(e.target.value)}
          role="textbox"
          aria-label="Search products"
        />
        {isLoading && <div data-testid="search-loading" aria-label="Loading">Loading...</div>}
      </div>
    );
  };
});

// Increase test timeout
jest.setTimeout(10000);

import useProducts from '../hooks/useProducts';

describe('ProductList Component', () => {
  const mockProducts = [
    { id: 1, name: 'Product 1', price: 10 },
    { id: 2, name: 'Product 2', price: 20 },
    { id: 3, name: 'Product 3', price: 30 }
  ];

  beforeEach(() => {
    useProducts.mockClear();
  });

  test('displays loading state when products are being fetched', () => {
    useProducts.mockReturnValue({
      products: [],
      loading: true,
      error: null
    });

    render(<ProductList onAddToCart={() => {}} />);
    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });

  test('displays error message when there is an error', () => {
    const errorMessage = 'Failed to fetch products';
    useProducts.mockReturnValue({
      products: [],
      loading: false,
      error: errorMessage
    });

    render(<ProductList onAddToCart={() => {}} />);
    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });

  test('displays "No products found" message when product list is empty', () => {
    useProducts.mockReturnValue({
      products: [],
      loading: false,
      error: null
    });

    render(<ProductList onAddToCart={() => {}} />);
    expect(screen.getByText('No products found.')).toBeInTheDocument();
  });

  test('renders grid layout with product cards when products are available', async () => {
    useProducts.mockReturnValue({
      products: mockProducts,
      loading: false,
      error: null
    });

    render(
      <ThemeProvider theme={theme}>
        <ProductList onAddToCart={() => {}} />
      </ThemeProvider>
    );

    // Wait for grid layout to be rendered
    await waitFor(() => {
      // Check if search bar is rendered
      expect(screen.getByTestId('search-bar')).toBeInTheDocument();

      // Check if all product cards are rendered
      mockProducts.forEach(product => {
        expect(screen.getByTestId(`product-card-${product.id}`)).toBeInTheDocument();
        expect(screen.getByText(product.name)).toBeInTheDocument();
      });

      // Check if grid container is present
      const gridContainer = document.querySelector('.MuiGrid-container');
      expect(gridContainer).toBeInTheDocument();

      // Check if grid items are present
      const gridItems = document.querySelectorAll('.MuiGrid-item');
      expect(gridItems).toHaveLength(mockProducts.length);
    });
  });

  describe('Snapshot Tests', () => {
    const renderWithTheme = async (component) => {
      let tree;
      await act(async () => {
        tree = renderer.create(
          <ThemeProvider theme={theme}>
            {component}
          </ThemeProvider>
        );
        await waitForMuiAnimation();
      });
      return tree;
    };

    test('matches loading state snapshot', async () => {
      useProducts.mockReturnValue({
        products: [],
        loading: true,
        error: null
      });

      const tree = await renderWithTheme(<ProductList onAddToCart={() => {}} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

    test('matches error state snapshot', async () => {
      const errorMessage = 'Failed to fetch products';
      useProducts.mockReturnValue({
        products: [],
        loading: false,
        error: errorMessage
      });

      const tree = await renderWithTheme(<ProductList onAddToCart={() => {}} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

    test('matches empty products state snapshot', async () => {
      useProducts.mockReturnValue({
        products: [],
        loading: false,
        error: null
      });

      const tree = await renderWithTheme(<ProductList onAddToCart={() => {}} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

    test('matches populated products list snapshot', async () => {
      useProducts.mockReturnValue({
        products: mockProducts,
        loading: false,
        error: null
      });

      const tree = await renderWithTheme(<ProductList onAddToCart={() => {}} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

    test('matches snapshot with different screen sizes', async () => {
      useProducts.mockReturnValue({
        products: mockProducts,
        loading: false,
        error: null
      });

      const originalMatchMedia = window.matchMedia;
      const mockMatchMedia = (query) => ({
        matches: query === '(min-width: 960px)',
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      });

      window.matchMedia = mockMatchMedia;
      const desktopTree = await renderWithTheme(<ProductList onAddToCart={() => {}} />);
      expect(desktopTree.toJSON()).toMatchSnapshot('desktop view');

      window.matchMedia = (query) => ({
        ...mockMatchMedia(query),
        matches: false,
      });
      const mobileTree = await renderWithTheme(<ProductList onAddToCart={() => {}} />);
      expect(mobileTree.toJSON()).toMatchSnapshot('mobile view');

      window.matchMedia = originalMatchMedia;
    });
  });

  describe('Search Integration', () => {
    const searchResults = [
      { id: 1, name: 'Laptop', price: 999 },
      { id: 2, name: 'Laptop Bag', price: 49 }
    ];

    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.useRealTimers();
    });

    test('updates search results when search query changes', async () => {
      // Initial render with empty results
      useProducts.mockReturnValue({
        products: [],
        loading: false,
        error: null
      });

      render(<ProductList onAddToCart={() => {}} />);

      // Mock the search results
      useProducts.mockReturnValue({
        products: searchResults,
        loading: false,
        error: null
      });

      // Trigger search
      fireEvent.change(screen.getByTestId('search-input'), {
        target: { value: 'laptop' }
      });

      // Fast-forward timers to handle debounce
      act(() => {
        jest.advanceTimersByTime(300);
      });

      // Wait for the results to be displayed
      await waitFor(() => {
        searchResults.forEach(product => {
          expect(screen.getByTestId(`product-card-${product.id}`)).toBeInTheDocument();
          expect(screen.getByText(product.name)).toBeInTheDocument();
        });
      });
    });

    test('shows loading state during search', async () => {
      // Initial state
      useProducts.mockReturnValue({
        products: [],
        loading: true,
        error: null
      });

      render(<ProductList onAddToCart={() => {}} />);

      // Trigger search
      fireEvent.change(screen.getByTestId('search-input'), {
        target: { value: 'laptop' }
      });

      // Verify loading state
      expect(screen.getByTestId('search-loading')).toBeInTheDocument();

      // Update mock to finished state
      useProducts.mockReturnValue({
        products: searchResults,
        loading: false,
        error: null
      });

      // Fast-forward timers
      act(() => {
        jest.advanceTimersByTime(300);
      });

      // Verify loading indicator is removed
      await waitFor(() => {
        expect(screen.queryByTestId('search-loading')).not.toBeInTheDocument();
      });
    });

    test('handles empty search results', async () => {
      // Initial render
      useProducts.mockReturnValue({
        products: mockProducts,
        loading: false,
        error: null
      });

      render(<ProductList onAddToCart={() => {}} />);

      // Update to empty results
      useProducts.mockReturnValue({
        products: [],
        loading: false,
        error: null
      });

      // Trigger search
      fireEvent.change(screen.getByTestId('search-input'), {
        target: { value: 'nonexistent' }
      });

      // Fast-forward timers
      act(() => {
        jest.advanceTimersByTime(300);
      });

      // Verify "No products found" message
      await waitFor(() => {
        expect(screen.getByText('No products found.')).toBeInTheDocument();
      });
    });

    test('handles search error state', async () => {
      const errorMessage = 'Search failed';
      
      // Initial render
      useProducts.mockReturnValue({
        products: mockProducts,
        loading: false,
        error: null
      });

      render(<ProductList onAddToCart={() => {}} />);

      // Update to error state
      useProducts.mockReturnValue({
        products: [],
        loading: false,
        error: errorMessage
      });

      // Trigger search
      fireEvent.change(screen.getByTestId('search-input'), {
        target: { value: 'error' }
      });

      // Fast-forward timers
      act(() => {
        jest.advanceTimersByTime(300);
      });

      // Verify error message
      await waitFor(() => {
        expect(screen.getByText(errorMessage)).toBeInTheDocument();
      });
    });
  });
});
