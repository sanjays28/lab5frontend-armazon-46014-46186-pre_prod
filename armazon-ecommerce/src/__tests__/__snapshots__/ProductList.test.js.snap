// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList Component Snapshot Tests matches empty products state snapshot 1`] = `
<div
  className="MuiBox-root css-1aicb05"
>
  <p
    className="MuiTypography-root MuiTypography-body1 css-10utshc-MuiTypography-root"
    data-testid="no-products-message"
  >
    No products found.
  </p>
</div>
`;

exports[`ProductList Component Snapshot Tests matches error state snapshot 1`] = `
<div
  className="MuiBox-root css-1aicb05"
>
  <p
    className="MuiTypography-root MuiTypography-body1 css-5okc32-MuiTypography-root"
    data-testid="error-message"
  >
    Failed to fetch products
  </p>
</div>
`;

exports[`ProductList Component Snapshot Tests matches loading state snapshot 1`] = `
<div
  className="MuiBox-root css-zzzt33"
>
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
    data-testid="loading-spinner"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </span>
</div>
`;

exports[`ProductList Component Snapshot Tests matches populated products list snapshot 1`] = `
<div
  className="MuiBox-root css-10wxp4g"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      className="search-bar"
      data-testid="search-bar"
    >
      <input
        aria-label="Search products"
        data-testid="search-input"
        onChange={[Function]}
        placeholder="Search products..."
        role="textbox"
        type="text"
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
      >
        <div
          className="product-card"
          data-testid="product-card-1"
        >
          <div
            data-testid="product-name-1"
          >
            Product 1
          </div>
          <div
            data-testid="product-price-1"
          >
            $
            10
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
      >
        <div
          className="product-card"
          data-testid="product-card-2"
        >
          <div
            data-testid="product-name-2"
          >
            Product 2
          </div>
          <div
            data-testid="product-price-2"
          >
            $
            20
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
      >
        <div
          className="product-card"
          data-testid="product-card-3"
        >
          <div
            data-testid="product-name-3"
          >
            Product 3
          </div>
          <div
            data-testid="product-price-3"
          >
            $
            30
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProductList Component Snapshot Tests matches snapshot with different screen sizes: desktop view 1`] = `
<div
  className="MuiBox-root css-10wxp4g"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      className="search-bar"
      data-testid="search-bar"
    >
      <input
        aria-label="Search products"
        data-testid="search-input"
        onChange={[Function]}
        placeholder="Search products..."
        role="textbox"
        type="text"
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
      >
        <div
          className="product-card"
          data-testid="product-card-1"
        >
          <div
            data-testid="product-name-1"
          >
            Product 1
          </div>
          <div
            data-testid="product-price-1"
          >
            $
            10
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
      >
        <div
          className="product-card"
          data-testid="product-card-2"
        >
          <div
            data-testid="product-name-2"
          >
            Product 2
          </div>
          <div
            data-testid="product-price-2"
          >
            $
            20
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
      >
        <div
          className="product-card"
          data-testid="product-card-3"
        >
          <div
            data-testid="product-name-3"
          >
            Product 3
          </div>
          <div
            data-testid="product-price-3"
          >
            $
            30
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProductList Component Snapshot Tests matches snapshot with different screen sizes: mobile view 1`] = `
<div
  className="MuiBox-root css-10wxp4g"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      className="search-bar"
      data-testid="search-bar"
    >
      <input
        aria-label="Search products"
        data-testid="search-input"
        onChange={[Function]}
        placeholder="Search products..."
        role="textbox"
        type="text"
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
      >
        <div
          className="product-card"
          data-testid="product-card-1"
        >
          <div
            data-testid="product-name-1"
          >
            Product 1
          </div>
          <div
            data-testid="product-price-1"
          >
            $
            10
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
      >
        <div
          className="product-card"
          data-testid="product-card-2"
        >
          <div
            data-testid="product-name-2"
          >
            Product 2
          </div>
          <div
            data-testid="product-price-2"
          >
            $
            20
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
      >
        <div
          className="product-card"
          data-testid="product-card-3"
        >
          <div
            data-testid="product-name-3"
          >
            Product 3
          </div>
          <div
            data-testid="product-price-3"
          >
            $
            30
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
